service/gcs/core/gcs/storage.go

```diff
@@ -253,25 +253,34 @@ func (c *gcsCore) unmountMappedVirtualDisks(disks []prot.MappedVirtualDisk) erro
 	return nil
 }
 
-// mountMappedDirectories mounts the given mapped directories using a Plan9
+// mountMappedDirectory mounts the given mapped directory using a Plan9
 // filesystem with the given options.
-func (c *gcsCore) mountMappedDirectories(dirs []prot.MappedDirectory) error {
-	for _, dir := range dirs {
-		if !dir.CreateInUtilityVM {
-			return errors.New("we do not currently support mapping directories inside the container namespace")
-		}
-		if err := c.OS.MkdirAll(dir.ContainerPath, 0700); err != nil {
-			return errors.Wrapf(err, "failed to create directory for mapped directory %s", dir.ContainerPath)
-		}
-		var mountOptions uintptr
-		data := fmt.Sprintf("trans=vsock,port=%d", dir.Port)
-		if dir.ReadOnly {
-			mountOptions |= syscall.MS_RDONLY
-			data += ",noload"
-		}
-		if err := c.OS.Mount(dir.ContainerPath, dir.ContainerPath, "9p", mountOptions, data); err != nil {
-			return errors.Wrapf(err, "failed to mount directory for mapped directory %s", dir.ContainerPath)
-		}
+func (c *gcsCore) mountMappedDirectory(dir *prot.MappedDirectory) error {
+	if !dir.CreateInUtilityVM {
+		return errors.New("we do not currently support mapping directories inside the container namespace")
+	}
+	if err := c.OS.MkdirAll(dir.ContainerPath, 0700); err != nil {
+		return errors.Wrapf(err, "failed to create directory for mapped directory %s", dir.ContainerPath)
+	}
+	conn, err := c.vsock.Dial(dir.Port)
```

Comment:
This change eliminates the need for the 9pfs patch, which is good.

The only question I have is: are there any cleanup need to be done in unmountMappedVirtualDisks when this mapping is no longer needed?

Reply:
Other than that, the change looks good to me.

Reply:
No, there's no other cleanup necessary. What happens is that the kernel references the file descriptor as part of the mount call, so after the mount call returns we can close the fds immediately. Then after unmount the kernel will release its reference and the socket will close without any additional code in GCS.